Graphs

capttrousers opened the pull request:
this implements the graph classes

capttrousers committed:
Adds node, edge, graph interfaces and implementing classes, closes #14 #15 #16

capttrousers committed:
adds classes missed in last commit

shalka commented:
I like where this is headed. Can you implement the UnweightedEdge class and include a getValue() in the Node interface (and implement it on GraphNode)? It should really be templated, but if you're not comfortable with that yet, just give it a type like int and we'll keep moving.

capttrousers committed:
updates node to have value / object and adds unweighted edge class

capttrousers commented:
i added the needful, what do you mean by templated? I just had it return a generic Object for now, because if the value is a primitive, we can just store it in the primitive wrapper object.

shalka committed:
make Node interface accept a generic type for value

shalka commented:
My bad. Spending too much time in c++ land lately. I meant [Generics](https://docs.oracle.com/javase/tutorial/java/generics/types.html). I've just pushed a change that illustrates what I mean, so that now we can put any value in any class that implements Node, like Node<Integer> or Node<String>.

capttrousers committed:
adds graph search class and begins DFS implementation

capttrousers committed:
finishes? DFS implementation

capttrousers committed:
fixes loop logic - DFS incremental implementation

shalka commented on src/main/java/graphs/Edge.java:
```diff
@@ -0,0 +1,8 @@
+    Node getLeftNode();
+    Node getRightNode();
+    boolean contains(Node node);
+    Node getOtherNode(Node node);
```
what's the distinction between getLeft, getRight, and getOther?

shalka commented on src/main/java/graphs/GraphSearch.java:
```diff
@@ -0,0 +1,77 @@
+        // DFS uses a stack of nodes visited
+        // stack is push pop peek : peek to see
+        LinkedList<Node> locationStack = new LinkedList<Node>();
+        Node startingNode = ((Node)graph.getNodes().toArray()[0]);
```
I have a problem with this because getNodes() returns a Set, not a List, so you're not guaranteed any ordering of Nodes. The depthFirstSearch method should take two additional arguments, Node start and Node end, that represent the start and end of the search criteria through the Graph.

shalka committed:
add visit() and unvisit() methods to GraphNode

shalka committed:
add test cases and a bunch of other tweaks

capttrousers committed:
passes 3 of 4 tests

capttrousers committed:
passes 3 of 4 tests

capttrousers commented:
so far it fails one test because it's not BFS, it misses the shorter route to H thru Z

capttrousers committed:
implements BFS, passes 3 of 4 tests

capttrousers commented:
so now I implemented BFS and changed the tests to use the appropriate algo, BFS test 1 passes, but the one with the shorter path to Z fails, with an error about not being able to add Node to the PriorityQueue since Node isn't a Comparable.  So I am thinking maybe there is a basic Queue like the Stack I used in DFS, or maybe I should just make that a LinkedList, as that will still retain FIFO which is what I need to track the Nodes to traverse.

capttrousers committed:
adds two more test graphs

capttrousers committed:
fixes BFS with transitions list to get path home, and DFS to return stack

capttrousers committed:
implements recursive DFS

capttrousers committed:
tests: fully connected graph, empty/null graph, null start/finish nodes fix #20
